src/util/pcp.cpp

```diff
@@ -0,0 +1,524 @@
+            }
+            if (response[0] != NATPMP_VERSION || response[1] != (NATPMP_RESPONSE | NATPMP_OP_MAP_TCP)) {
+                LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "natpmp: Response to wrong command\n");
+                return false; // Wasn't response to what we expected, try receiving next packet.
+            }
+            uint16_t internal_port = ReadBE16(response.data() + NATPMP_MAP_RESPONSE_INTERNAL_PORT_OFS);
+            if (internal_port != port) {
+                LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "natpmp: Response port doesn't match request\n");
+                return false; // Wasn't response to what we expected, try receiving next packet.
+            }
+            return true;
+        });
+
+    if (recv_res) {
+        const std::span<uint8_t> response = *recv_res;
+
+        Assume(response.size() >= NATPMP_MAP_RESPONSE_SIZE);
+        uint16_t result_code = ReadBE16(response.data() + NATPMP_RESPONSE_HDR_RESULT_OFS);
+        if (result_code != NATPMP_RESULT_SUCCESS) {
+            LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "natpmp: Port mapping failed with result %s\n", NATPMPResultString(result_code));
+            if (result_code == NATPMP_RESULT_NO_RESOURCES) {
+                return MappingError::NO_RESOURCES;
+            }
+            return MappingError::PROTOCOL_ERROR;
+        }
+
+        uint32_t lifetime_ret = ReadBE32(response.data() + NATPMP_MAP_RESPONSE_LIFETIME_OFS);
+        uint16_t external_port = ReadBE16(response.data() + NATPMP_MAP_RESPONSE_EXTERNAL_PORT_OFS);
+        return MappingResult(NATPMP_VERSION, CService(internal.sin_addr, port), CService(external_addr, external_port), lifetime_ret);
+    } else {
+        return MappingError::NETWORK_ERROR;
+    }
+}
+
+std::variant<MappingResult, MappingError> PCPRequestPortMap(const PCPMappingNonce &nonce, const CNetAddr &gateway, const CNetAddr &bind, uint16_t port, uint32_t lifetime, int num_tries, std::chrono::milliseconds timeout_per_try)
+{
+    struct sockaddr_storage dest_addr, bind_addr;
+    socklen_t dest_addrlen = sizeof(struct sockaddr_storage), bind_addrlen = sizeof(struct sockaddr_storage);
+
+    LogPrintLevel(BCLog::NET, BCLog::Level::Debug, "pcp: Requesting port mapping for addr %s port %d from gateway %s\n", bind.ToStringAddr(), port, gateway.ToStringAddr());
+
+    // Validate addresses, make sure they're the same network family.
+    if (!CService(gateway, PCP_SERVER_PORT).GetSockAddr((struct sockaddr*)&dest_addr, &dest_addrlen)) return MappingError::NETWORK_ERROR;
+    if (!CService(bind, 0).GetSockAddr((struct sockaddr*)&bind_addr, &bind_addrlen)) return MappingError::NETWORK_ERROR;
+    if (dest_addr.ss_family != bind_addr.ss_family) return MappingError::NETWORK_ERROR;
+
+    // Create UDP socket (IPv4 or IPv6 based on provided gateway).
+    auto sock{CreateSock(dest_addr.ss_family, SOCK_DGRAM, IPPROTO_UDP)};
```

Comment:
I guess this is acceptable, but just mentioning: my understanding of RFC 6887 is that the server can send unsolicited messages with updates at any time, in case of address changes. When `PCPRequestPortMap()` ends, then this `sock` will go out of scope, will be closed and we will miss such replies.

Reply:
i don't think that's the case in practice? but even so, doing this differently would really complicate things too much, changes in network configuration would be noticed the next update interval anyway?

Reply:
Yes, missed updates should result in us advertising the wrong (old) address only until the next renewal.